facebook/delphiFacebook/R/contingency_tables.R

```diff
@@ -0,0 +1,1068 @@
+    {
+      s_mix_coef <- params$s_mix_coef
+      mixing <- mix_weights(agg_df[[var_weight]] * agg_df$weight_in_location,
+                            s_mix_coef, params$s_weight)
+      
+      sample_size <- sum(agg_df$weight_in_location)
+      
+      ## TODO Fix this. Old pipeline for community responses did not apply
+      ## mixing. To reproduce it, we ignore the mixed weights. Once a better
+      ## mixing/weighting scheme is chosen, all signals should use it.
+      new_row <- compute_fn(
+        response = agg_df[[metric]],
+        weight = if (aggregations$skip_mixing[row]) { mixing$normalized_preweights } else { mixing$weights },
+        sample_size = sample_size)
+      
+      dfs_out[[aggregation]]$val <- new_row$val
+      dfs_out[[aggregation]]$se <- new_row$se
+      dfs_out[[aggregation]]$sample_size <- sample_size
+      dfs_out[[aggregation]]$effective_sample_size <- new_row$effective_sample_size
+    }
+  }
+  
+  return(dfs_out)
+}
+
+
+#' Write csv file for sharing with researchers
+#' 
+#' CSV name includes date specifying start of time period aggregated, geo level,
+#' and grouping variables.
+#'
+#' @param data           a data frame to save; must contain the columns "geo_id", "val",
+#'                       "se", "sample_size", and grouping variables. The first four are saved in the
+#'                       output; day is used for spliting the data into files.
+#' @param params         a named list, containing the value "export_dir" indicating the
+#'                       directory where the csv should be saved
+#' @param geo_name       name of the geographic level; used for naming the output file
+#' @param signal_name    name of the signal; used for naming the output file
+#'
+#' @importFrom readr write_csv
+#' @importFrom dplyr arrange_at
+#' 
+#' @export
+write_contingency_tables <- function(data, params, geo_level, groupby_vars)
+{
+  if (!is.null(data) && nrow(data) != 0) {
+    data <- arrange_at(data, groupby_vars)
+  } else {
+    msg_plain(sprintf(
+      "no aggregations produced for grouping variables %s (%s); CSV will not be saved", 
+      paste(groupby_vars, collapse=", "), geo_level
+    ))
+    return()
+  }
+  
+  file_out <- file.path(
+    params$export_dir, sprintf("%s_%s_%s.csv", format(params$start_date, "%Y%m%d"),
+                               geo_level, paste(groupby_vars, collapse="_"))
+  )
+  
+  create_dir_not_exist(params$export_dir)
+  
+  msg_df(sprintf(
+    "saving contingency table data to %-35s",
+    sprintf("%s_%s_%s", format(params$start_date, "%Y%m%d"),
+            geo_level, paste(groupby_vars, collapse="_"))
+  ), data)
+  write_csv(data, file_out)
+}
+
+
+
+#' Get the date of the first day of the previous month
+#'
+#' @param date Date that output will be calculated relative to
+#' 
+#' @return Date
+#' 
+#' @importFrom lubridate floor_date
+#' 
+#' @export
+start_of_prev_full_month <- function(date) {
+  return(floor_date(date, "month") - months(1))
+}
```

Comment:
Maybe these date util functions can go into a `dates.R` or something? Just to keep the files manageable in size

Reply:
Yes! I initially grouped the functions for convenience and haven't thought too much about splitting them up sensibly. Since a lot of them are derivatives of existing FB package functions, I could use the same file structure but prefix with `contingency_tables_*`, so `contingency_tables_aggregate.R`, etc. Do you have a particular preference here?

Reply:
I'd prefix with `contingency` rather than `contingency_tables` (just easier to deal with), but that sounds good. I think these date functions aren't contingency-specific, though, even though the tables are their only current use case.